packages/react-utilities/src/compose/types.ts

```diff
@@ -16,17 +16,64 @@ export type ShorthandProps<Props extends DefaultObjectShorthandProps> =
   | undefined // default render (or null render if no default provided)
   | Props;
 
-export type DefaultObjectShorthandProps = ObjectShorthandProps<{}, unknown, keyof JSX.IntrinsicElements>;
-
-export type ObjectShorthandProps<
-  Props extends { children?: React.ReactNode } = {},
-  Ref = unknown,
-  As extends keyof JSX.IntrinsicElements = never
-> = Props &
-  React.RefAttributes<Ref> & {
-    as?: As;
-    children?: Props['children'] | ShorthandRenderFunction<Props>;
-  };
+/**
+ * Matches any shorthand props type.
+ *
+ * This should ONLY be used in type templates as in `extends DefaultObjectShorthandProps`;
+ * it shouldn't be used as the type of a slot.
+ */
+export type DefaultObjectShorthandProps = ObjectShorthandProps<{
+  children?: React.ReactNode;
+  as?: keyof JSX.IntrinsicElements;
+}>;
+
+/**
+ * Defines the slot props for a slot that supports a Component type.
+ *
+ * For intrinsic elements like 'div', use {@link ElementShorthandProps} instead.
+ */
+export type ObjectShorthandProps<Props extends { children?: React.ReactNode } = {}> = Props & {
+  children?: Props['children'] | ShorthandRenderFunction<Props>;
+};
+
+/**
+ * Define the slot arguments for a slot that supports one or more intrinsic element types, such as 'div'.
+ * For slots that support custom components, use {@link ObjectShorthandProps} instead.
+ *
+ * The first param is the slot's default type if no `as` prop is specified.
+ * The second param is an optional union of alternative types that can be specified for the `as` prop.
+ *
+ * ```
+ * ElementShorthandProps<'div'> // Slot is always a <div>
+ * ElementShorthandProps<'a', 'button'> // Defaults to <a>, but allows { as: 'button' } to be a <button>
+ * ElementShorthandProps<'label', 'span' | 'div'>; // Defaults to <label>, but allows { as: 'span' } or { as: 'div' }
+ * ```
+ */
+export type ElementShorthandProps<
+  DefaultElement extends keyof JSX.IntrinsicElements,
+  AlternateElements extends keyof JSX.IntrinsicElements = never
+> = IsSingleton<DefaultElement> extends false
+  ? 'Error: first parameter to ElementShorthandProps must be a single element type, not a union of types'
+  :
+      | ({ as?: DefaultElement } & ObjectShorthandProps<NoLegacyRef<JSX.IntrinsicElements[DefaultElement]>>)
+      | {
+          [As in AlternateElements]: { as: As } & ObjectShorthandProps<NoLegacyRef<JSX.IntrinsicElements[As]>>;
+        }[AlternateElements];
+
+/**
+ * Evaluates to true if the given type contains exactly one string, or false if it is a union of strings.
+ *
+ * ```
+ * IsSingleton<'a'> // true
+ * IsSingleton<'a' | 'b' | 'c'> // false
+ * ```
+ */
+export type IsSingleton<T extends string> = { [K in T]: Exclude<T, K> extends never ? true : false }[T];
+
+/**
+ * Excludes LegacyRef (string) from the ref prop of the given Props type
+ */
+export type NoLegacyRef<Props extends { ref?: unknown }> = Omit<Props, 'ref'> & { ref?: Exclude<Props['ref'], string> };
```

Comment:
Is this type really necessary ? If it's still in the official typings let's keep it

Reply:
Yes, it's necessary to compile. Without this, all of our component's Props types would allow legacy refs, except none of our component code is written to handle legacy refs. I think `React.forwardRef` also doesn't allow legacy refs.

(When `ElementShorthandProps` is used for the `root` slot, then the `ref` prop from here gets mixed in with the component's Props type.)